packages/nextcloud/lib/src/webdav/client.dart

```diff
@@ -75,6 +80,24 @@ class WebDavClient {
       ...?rootClient.authentications?.firstOrNull?.headers,
     });
 
+    if (_kIsWeb) {
+      if (_token == null) {
+        final response = await rootClient.httpClient.get(Uri.parse('${rootClient.baseURL}/index.php'));
+        if (response.statusCode >= 300) {
+          throw DynamiteStatusCodeException(
+            response.statusCode,
+          );
+        }
+
+        _token = RegExp('data-requesttoken="([^"]*)"').firstMatch(response.body)!.group(1);
+      }
+
+      request.headers.addAll({
+        'OCS-APIRequest': 'true',
+        'requesttoken': _token!,
+      });
+    }
+
```

Comment:
Short answer: We have to do this because of cookies. On other platforms we don't send the cookies so it is not a problem. The browser always sends the cookies, so that forces us to have this work around.
Long answer: This is probably a bug in the server code and we should be required to always add this token, but it doesn't happen. I'm going to do some more digging into that.

The token is indeed cached and can be reused, but I don't know for how long.

Reply:
Then we should find a better solution for this looks quite expensive.
https://github.com/nextcloud/neon/issues/1536 together with  #1580 will also send cookies on non web platforms.

Reply:
I had a look at the code and the csrf tokens are bound to the php session. That means the token can be invalid at any point in time, so we would need a proper logic for catching that, getting a new token and retrying the request.
I will have a look at the server code with some coworkers because I think it has some bugs. Maybe we can also figure out if it even makes sense to have this csrf token, since it just adds another hurdle without preventing anyone from actually using webdav afaict :thinking:

Reply:
So there are definitely bugs and it should just work for us when we send the OCS-APIRequest, but it doesn't.
Would you be fine with merging this as a temporary workaround (with explanation comments in the code) until it is fixed in the server?

Reply:
Sure; There is nothing we can do.
I'll try to give the final thumbs up review this evening but as I already had a look multiple times there shouldn't be any real blocker.

Reply:
Ok, I'll try to add the comments soon.